### PR TITLE
[Clang][OpenMP] Fix ordering of processing of map clauses when mapping a struct.

### PR DESCRIPTION
**Pull request by doru1004**
Mapping a struct, if done in the wrong order, can overwrite the pointer attachment details. This fixes this problem.

Original failing example:

```
#include <stdio.h>
#include <stdlib.h>

struct Descriptor {
  int *datum;
  long int x;
  int xi;
  long int arr[1][30];
};

int main() {
  Descriptor dat = Descriptor();
  dat.datum = (int *)malloc(sizeof(int)*10);
  dat.xi = 3;
  dat.arr[0][0] = 1;

  #pragma omp target enter data map(to: dat.datum[:10]) map(to: dat)

  #pragma omp target
  {
    dat.xi = 4;
    dat.datum[dat.arr[0][0]] = dat.xi;
  }

  #pragma omp target exit data map(from: dat)

 return 0;
}
```

Previous attempt at fixing this: https://github.com/llvm/llvm-project/pull/70821

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
This being in clang instead seems like a good change. Are there no CodeGen  tests changed? We should add one if so.  Probably just take your `libomptarget` test and run `update_cc_test_checks` on it with the arguments found in other test files.

**Comment by doru1004** [comment]
> This being in clang instead seems like a good change. Are there no CodeGen tests changed? We should add one if so. Probably just take your `libomptarget` test and run `update_cc_test_checks` on it with the arguments found in other test files.

No code gen test changes. Happy to add one no problem.

**Comment by doru1004** [comment]
> This being in clang instead seems like a good change. Are there no CodeGen tests changed? We should add one if so. Probably just take your `libomptarget` test and run `update_cc_test_checks` on it with the arguments found in other test files.

Just added the test.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 1]
```diff
@@ -7742,15 +7744,42 @@ class MappableExprsHandler {
...
+      const auto *EI = C->getVarRefs().begin();
+      if (*EI && isa<OMPArraySectionExpr>(*EI)) {
+        for (const auto L : C->component_lists()) {
+          const Expr *E = (C->getMapLoc().isValid()) ? *EI : nullptr;
+          InfoGen(std::get<0>(L), Kind, std::get<1>(L), C->getMapType(),
+                  C->getMapTypeModifiers(), std::nullopt,
+                  /*ReturnDevicePointer=*/false, C->isImplicit(),
+                  std::get<2>(L), E);
+          ++EI;
+        }
```
Comment (jdoerfert) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:7779]: This duplicates the loop nest, which is very unfortunate. Why not actually sort the clause list? That will also make it easier to add/change things in the future, e.g., we simply modify the comparator.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 2]
```diff
@@ -7742,15 +7744,42 @@ class MappableExprsHandler {
       else if (C->getMapType() == OMPC_MAP_alloc)
         Kind = Allocs;
       const auto *EI = C->getVarRefs().begin();
-      for (const auto L : C->component_lists()) {
-        const Expr *E = (C->getMapLoc().isValid()) ? *EI : nullptr;
-        InfoGen(std::get<0>(L), Kind, std::get<1>(L), C->getMapType(),
-                C->getMapTypeModifiers(), std::nullopt,
-                /*ReturnDevicePointer=*/false, C->isImplicit(), std::get<2>(L),
-                E);
-        ++EI;
```
Comment (alexey-bataev) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:7751]: That's strange that array sections are not handled correctly here, need to check why componen_list() does not return it correctly. Or just InfoGen does not process it correctly.

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 3]
```diff
@@ -7731,10 +7731,30 @@ class MappableExprsHandler {
               IsImplicit, Mapper, VarRef, ForDeviceAddr);
         };
 
+    // Sort all map clauses and make sure all the maps containing array
+    // sections are processed last.
+    llvm::SmallVector<const OMPMapClause *, 16> SortedMapClauses;
```
Comment (jhuber6) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: What implies that this is sorted?

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: I don't understand the question.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: Are you asking what is the sorting criteria?

Comment (jhuber6) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: It's called "SortedMapClauses" but I don't see any sorting, we just push back into the vector.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: Ah yes, so I just moved all the maps containing sections at the end of the clause list. I want those maps to happen last after all the structs and other maps have happened.

Comment (alexey-bataev) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: This sorting approach is not very good, it adds lots of the code which is not required. It was supposed that this function automatically adjusts the boundaries of the structs/classes, so sorting is not required.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: It's a form of sorting it's more like a split between all section-containing maps and the ones that don't.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: @alexey-bataev I agree it's not ideal. the problem is related to the order in which the clauses are processed. We cannot process the base struct after we have processed an array section inside the struct.

Comment (alexey-bataev) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: Why, what's wrong with this?

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: at runtime, if things happen in the wrong order, the processing of the base struct overwrites the pointer attachment for the array.

Comment (alexey-bataev) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: But that what this code already does, it (implicitly) "reorders" the clauses by extending the mapped region automatically. Explicit reordering should not be required.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: Implicit ordering isn't working in the case in the example above, please see the code. The entries are in the wrong order in the runtime and the problem starts here.

Comment (alexey-bataev) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: And this is what should be fixed, no need to add a hack, that "hides" the bug in the code, need to fix the bug.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: Well I don't see anything other that's wrong other than the order and the order comes from how the user wrote the code so I am not sure how else to fix it.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: I can't find any "bug" in the existing code. It works as intended. The problem is that it doesn't handle these types of situations and I don't see how else to fix an ordering problem other than by re-ordering. If you have a different solution in mind please let me know.

Comment (doru1004) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: @alexey-bataev I have looked at the code again and I really can't see another solution to this problem. If you have a different fix in mind please let me know.

Comment (alexey-bataev) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:7736]: Will try to investigate it tomorrow, probably

**Pull request closed without merging**